File: PA_40.py
#LC 98. Validate Binary Search Tree
#Time complexity - O(n)
#Space Complexity - O(h)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        self.prev = None
        def inorder(root):
            if not root:
                return True
            
            if inorder(root.left) == False:
                return False
            
            if self.prev != None and root.val <= self.prev.val:
                return False
            self.prev = root
            return inorder(root.right)
        return inorder(root)

File: PA40.java
//Iteravtive solution
//Valid BST
//Time : O(n) , n is no of nodes in the tree
//Space : O(n) , where n is the height of the tree. In worst case(Skewed tree) log(n) height will be same as n.
class Solution {
    public boolean isValidBST(TreeNode root) {
        Stack<TreeNode> st = new Stack<>();
        
        TreeNode prev=null;
        while(!st.isEmpty() || root!=null){
            while(root!=null){
                
                st.push(root);
                root= root.left;
            }
            
            root = st.pop();
            
            if(prev!=null && prev.val>= root.val){
                return false;
            }
            prev = root;
            root=root.right;
            
        }
        return true;
    }
}

//====================================================
//Valid BST Recursive solution
//Time complexity : O(n) // we are itrating every node of tree
//Space Complexity: O(n) // It the space of stack used in recurssion to store tree nodes.
class Solution {
    TreeNode prev= null;
    public boolean isValidBST(TreeNode root) {
        if(root==null){
            return true;
        }
      return  helper(root);
    }
    public boolean helper(TreeNode root){
        
        if(root==null){
           
            return true;
        }
        if(!helper(root.left)) return false;
        if(prev!=null && prev.val>= root.val) return false;
        prev= root;
        
        return helper(root.right);
        
    }
}